### DISCUSSION TITLE
[Interpret model] SHAP values to data frame and custom top values

### DISCUSSION THREAD

**Discussion by victor-mariano-leite** [pycaret/pycaret#1351] [Q&A]
**Is your feature request related to a problem? Please describe.**

As far as I've seen, we can only plot the SHAP values when model is trained, it would be useful if we had an parameter to output to a dataframe, instead of a HTML visualization.

And set the ranking threshold of the visualization, that it's hardcoded as 25.

**Describe the solution you'd like**

output = 'html' (default) parameter, and output = 'df' to output the SHAP mean values as a dataframe, it would simplify the versioning and comparing the values and rankings with other model versions.

Implemented simply like this in [issue #970](https://github.com/slundberg/shap/issues/970) from SHAP.

top parameter, defaults to 25, on top of SHAP max_display parameter.

**Comment by ngupta23** [comment]
You can extract the shapley values manually for now. Refer to [this notebook](https://nbviewer.jupyter.org/github/ngupta23/ds7333_qtw/blob/master/case_study_7/submission_Kannan_Moro_Gupta/code/cs7.ipynb
) (it is towards the end)


```python
shap_values = shap.TreeExplainer(best_model).shap_values(X_train_transformed)
```

**Reply by alexxony** [reply]
I've known shap values are model-independant, but in my experiment, they were diffrent.

**Reply by alexxony** [reply]
or my metric is macro f1 score, in this case how can I handle??

**Comment by victor-mariano-leite** [comment]
@ngupta23 it's true! Nice, I'll close the issue then

**Comment by victorhwfreire** [comment]
The way @ngupta23 suggest works for models that don't have a probability_threshold and are not tuned.
I'm working on a model that is tuned and has a probability_threshold, then when I try to execute as mentioned I get the following error:
"Model type not yet supported by TreeExplainer: <class 'pycaret.internal.meta_estimators.CustomProbabilityThresholdClassifier'"
Does anyone have a solution for this?
Thanks

**Reply by ngupta23** [reply]
@victorhwfreire You can open up an issue here to see if they have a response on that: https://github.com/slundberg/shap
